### ISSUE TITLE
multiple torrents support

### ISSUE THREAD

**Issue by ThaUnknown** [jimmywarting/webtorrent-server-browser#1]
sorry for commenting on a semi-dead repo, but is there any way to make this work with multiple torrents at once? so far I've removed previous torrents and that worked fine but I wanted to seed multiple torrents at once and ran into issues

**Comment by jimmywarting** [comment]
> sorry for commenting on a semi-dead repo, 

No problem

> but is there any way to make this work with multiple torrents at once? so far I've removed previous torrents and that worked fine but I wanted to seed multiple torrents at once and ran into issues


What kind of problem did you run into?

**Comment by ThaUnknown** [comment]
> > sorry for commenting on a semi-dead repo,
> 
> No problem
> 
> > but is there any way to make this work with multiple torrents at once? so far I've removed previous torrents and that worked fine but I wanted to seed multiple torrents at once and ran into issues
> 
> What kind of problem did you run into?

mainly getmimetype of file, seems like the file is undefined, i dont think its because of my changes/fixes to this so i thought I'd ask here first, i could send the full error and code in about 1h

**Comment by ThaUnknown** [comment]
@jimmywarting okay adding 2 torrents: 
```addTorrent("https://webtorrent.io/torrents/big-buck-bunny.torrent")```
and then
```addTorrent("https://webtorrent.io/torrents/sintel.torrent")```
then requesting anything thats not yet downloaded from the 1st torrent gives
```
Uncaught TypeError: Cannot read property '_getMimeType' of undefined
    at serveFile (torrentHandler.js:132)
    at ServiceWorkerContainer.<anonymous> (torrentHandler.js:175)
  | serveFile | @ | torrentHandler.js:132
  | (anonymous) | @ | torrentHandler.js:175
```
my addTorrent code:
```js
function addTorrent(magnet) {
    // if (client.torrents[0]) {
    //     client.remove(client.torrents[0].infoHash)
    // }
    client.add(magnet, async function (torrent) {
        const server = await torrent.createServer()
        await server.listen()

        let videoFile = torrent.files[0]
        torrent.files.forEach(file => {
            if (file.length > videoFile.length) {
                videoFile = file
            }
        })
        video.src = `/webtorrent/${torrent.infoHash}/${encodeURI(videoFile.path)}`

    })
}
```

another change i made is in serviceworker's message event listener
```js
const file = torrent.files.find(f => f.path === decodeURIComponent(pathname))
``` 
because it had issues with some filenames

**Comment by ThaUnknown** [comment]
@jimmywarting any ideas?

**Comment by jimmywarting** [comment]
Hmm, have no real idea about what's going on, but decodeURI was a good idea that i have missed.
The code is kinda outdated i think. would have solved it in a diffrent way nowdays.

made a simpler version that seems to work:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>browser server test</title>
  </head>
  <body>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/webtorrent/0.108.6/webtorrent.min.js"></script>
    <script src="range-parser.js"></script>
    <script>
      const client = new WebTorrent()
      const sintel = 'magnet:?xt=urn:btih:08ada5a7a6183aae1e09d831df6748d566095a10&dn=Sintel&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969&tr=udp%3A%2F%2Ftracker.empire-js.us%3A1337&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337&tr=wss%3A%2F%2Ftracker.btorrent.xyz&tr=wss%3A%2F%2Ftracker.fastcast.nz&tr=wss%3A%2F%2Ftracker.openwebtorrent.com&ws=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2F'
      const bigBuck = 'https://webtorrent.io/torrents/big-buck-bunny.torrent'
      const scope = './'
      const sw = navigator.serviceWorker.register(`sw.js`, { scope })

      client.add(sintel, async function(torrent) {
        await sw
        const video = document.createElement('video')
        video.controls = true
        video.src = `./webtorrent/${torrent.infoHash}/${torrent.files[5].path}`
        video.style.width = '200px'
        document.body.appendChild(video)
      })

      client.add(bigBuck, async function(torrent) {
        await sw
        const video = document.createElement('video')
        video.controls = true
        video.src = `./webtorrent/${torrent.infoHash}/${torrent.files[1].path}`
        video.style.width = '200px'
        document.body.appendChild(video)
      })

      function serveFile (file, req) {
        const res = {
          status: 200,
          headers: {
            'Content-Type': file._getMimeType(),
            // Support range-requests
            'Accept-Ranges': 'bytes'
          }
        }

        // `rangeParser` returns an array of ranges, or an error code (number) if
        // there was an error parsing the range.
        let range = rangeParser(file.length, req.headers.get('range') || '')

        if (Array.isArray(range)) {
          res.status = 206 // indicates that range-request was understood

          // no support for multi-range request, just use the first range
          range = range[0]

          res.headers['Content-Range'] = `bytes ${range.start}-${range.end}/${file.length}`
          res.headers['Content-Length'] =  `${range.end - range.start + 1}`
        } else {
          range = null
          res.headers['Content-Length'] = file.length
        }

        res.body = req.method === 'HEAD' ? '' : 'stream'

        return [res, req.method === 'GET' && file.createReadStream(range)]
      }

      // kind of a fetch event from service worker but for the main thread.
      navigator.serviceWorker.addEventListener('message', evt => {
        const request = new Request(evt.data.url, {
          headers: evt.data.headers,
          method: evt.data.method
        })

        const [ port ] = evt.ports
        const respondWith = msg => port.postMessage(msg)
        const pathname = request.url.split(evt.data.scope + 'webtorrent/')[1]
        let [ infoHash, ...filePath ] = pathname.split('/')
        filePath = decodeURI(filePath.join('/'))

        if (!infoHash || !filePath) return

        const torrent = client.get(infoHash)
        const file = torrent.files.find(file => file.path === filePath)

        const [response, stream] = serveFile(file, request)
        const asyncIterator = stream && stream[Symbol.asyncIterator]()

        respondWith(response)

        async function pull () {
          respondWith((await asyncIterator.next()).value)
        }

        port.onmessage = pull
      })
    </script>
  </body>
</html>
```

```js
self.addEventListener('activate', evt =>  {
	return self.clients.claim()
})

self.addEventListener('fetch', evt => {
  const { request } = evt
  const { url, method, headers } = request
  // console.log(request)
  if (!url.includes(self.registration.scope + 'webtorrent/')) return null

  function getConsumer(clients) {
    return new Promise(rs => {
      // Use race condition for whoever controls the response stream
      for (const client of clients) {
        const mc = new MessageChannel()
        const {port1, port2} = mc
        port1.onmessage = evt => {
          rs([evt.data, mc])
        }
        client.postMessage({
          url,
          method,
          headers: [...headers],
          scope: self.registration.scope
        }, [port2])
      }
    })
  }

  evt.respondWith(
    clients.matchAll({ type: 'window', includeUncontrolled: true })
    .then(getConsumer)
    .then(([data, mc]) => {
      const body = data.body === 'stream' ? new ReadableStream({
        pull(controller) {
          return new Promise(rs => {
            mc.port1.onmessage = evt => {
              console.log(evt.data)
              evt.data
                ? controller.enqueue(evt.data) // evt.data is Uint8Array
                : controller.close() // evt.data is null, means the stream ended
              rs()
            }
            mc.port1.postMessage(true) // send a pull request
          })
        },
        cancel() {
          // This event is never executed
          console.log('request aborted')
          mc.port1.postMessage(false) // send a cancel request
        }
      }) : data.body

      return new Response(body, data)
    })
    .catch(console.error)
  )
})
```

**Comment by ThaUnknown** [comment]
> Hmm, have no real idea about what's going on, but decodeURI was a good idea that i have missed.
> The code is kinda outdated i think. would have solved it in a diffrent way nowdays.
> 
> made a simpler version that seems to work:
> 
> ```
> <!DOCTYPE html>
> <html>
>   <head>
>     <meta charset="utf-8">
>     <title>browser server test</title>
>   </head>
>   <body>
>     <script src="https://cdnjs.cloudflare.com/ajax/libs/webtorrent/0.108.6/webtorrent.min.js"></script>
>     <script src="range-parser.js"></script>
>     <script>
>       const client = new WebTorrent()
>       const sintel = 'magnet:?xt=urn:btih:08ada5a7a6183aae1e09d831df6748d566095a10&dn=Sintel&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969&tr=udp%3A%2F%2Ftracker.empire-js.us%3A1337&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337&tr=wss%3A%2F%2Ftracker.btorrent.xyz&tr=wss%3A%2F%2Ftracker.fastcast.nz&tr=wss%3A%2F%2Ftracker.openwebtorrent.com&ws=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2F'
>       const bigBuck = 'https://webtorrent.io/torrents/big-buck-bunny.torrent'
>       const scope = './'
>       const sw = navigator.serviceWorker.register(`sw.js`, { scope })
> 
>       client.add(sintel, async function(torrent) {
>         await sw
>         const video = document.createElement('video')
>         video.controls = true
>         video.src = `./webtorrent/${torrent.infoHash}/${torrent.files[5].path}`
>         video.style.width = '200px'
>         document.body.appendChild(video)
>       })
> 
>       client.add(bigBuck, async function(torrent) {
>         await sw
>         const video = document.createElement('video')
>         video.controls = true
>         video.src = `./webtorrent/${torrent.infoHash}/${torrent.files[1].path}`
>         video.style.width = '200px'
>         document.body.appendChild(video)
>       })
> 
>       function serveFile (file, req) {
>         const res = {
>           status: 200,
>           headers: {
>             'Content-Type': file._getMimeType(),
>             // Support range-requests
>             'Accept-Ranges': 'bytes'
>           }
>         }
> 
>         // `rangeParser` returns an array of ranges, or an error code (number) if
>         // there was an error parsing the range.
>         let range = rangeParser(file.length, req.headers.get('range') || '')
> 
>         if (Array.isArray(range)) {
>           res.status = 206 // indicates that range-request was understood
> 
>           // no support for multi-range request, just use the first range
>           range = range[0]
> 
>           res.headers['Content-Range'] = `bytes ${range.start}-${range.end}/${file.length}`
>           res.headers['Content-Length'] =  `${range.end - range.start + 1}`
>         } else {
>           range = null
>           res.headers['Content-Length'] = file.length
>         }
> 
>         res.body = req.method === 'HEAD' ? '' : 'stream'
> 
>         return [res, req.method === 'GET' && file.createReadStream(range)]
>       }
> 
>       // kind of a fetch event from service worker but for the main thread.
>       navigator.serviceWorker.addEventListener('message', evt => {
>         const request = new Request(evt.data.url, {
>           headers: evt.data.headers,
>           method: evt.data.method
>         })
> 
>         const [ port ] = evt.ports
>         const respondWith = msg => port.postMessage(msg)
>         const pathname = request.url.split(evt.data.scope + 'webtorrent/')[1]
>         let [ infoHash, ...filePath ] = pathname.split('/')
>         filePath = decodeURI(filePath.join('/'))
> 
>         if (!infoHash || !filePath) return
> 
>         const torrent = client.get(infoHash)
>         const file = torrent.files.find(file => file.path === filePath)
> 
>         const [response, stream] = serveFile(file, request)
>         const asyncIterator = stream && stream[Symbol.asyncIterator]()
> 
>         respondWith(response)
> 
>         async function pull () {
>           respondWith((await asyncIterator.next()).value)
>         }
> 
>         port.onmessage = pull
>       })
>     </script>
>   </body>
> </html>
> ```
> 
> ```js
> self.addEventListener('activate', evt =>  {
> 	return self.clients.claim()
> })
> 
> self.addEventListener('fetch', evt => {
>   const { request } = evt
>   const { url, method, headers } = request
>   // console.log(request)
>   if (!url.includes(self.registration.scope + 'webtorrent/')) return null
> 
>   function getConsumer(clients) {
>     return new Promise(rs => {
>       // Use race condition for whoever controls the response stream
>       for (const client of clients) {
>         const mc = new MessageChannel()
>         const {port1, port2} = mc
>         port1.onmessage = evt => {
>           rs([evt.data, mc])
>         }
>         client.postMessage({
>           url,
>           method,
>           headers: [...headers],
>           scope: self.registration.scope
>         }, [port2])
>       }
>     })
>   }
> 
>   evt.respondWith(
>     clients.matchAll({ type: 'window', includeUncontrolled: true })
>     .then(getConsumer)
>     .then(([data, mc]) => {
>       const body = data.body === 'stream' ? new ReadableStream({
>         pull(controller) {
>           return new Promise(rs => {
>             mc.port1.onmessage = evt => {
>               console.log(evt.data)
>               evt.data
>                 ? controller.enqueue(evt.data) // evt.data is Uint8Array
>                 : controller.close() // evt.data is null, means the stream ended
>               rs()
>             }
>             mc.port1.postMessage(true) // send a pull request
>           })
>         },
>         cancel() {
>           // This event is never executed
>           console.log('request aborted')
>           mc.port1.postMessage(false) // send a cancel request
>         }
>       }) : data.body
> 
>       return new Response(body, data)
>     })
>     .catch(console.error)
>   )
> })
> ```

major thanks for the help, i'll do some testing and try to make a PR if i find any issues, much love for this thing <3

**Comment by ThaUnknown** [comment]
seems like almost everything worked in this without issue

**Comment by ThaUnknown** [comment]
fixed in 1735d375381a9803f48f226334149b14236120b5
